### PR TITLE
[wasm] WasmHttpMessageHandler abort connection when response or respo…

### LINKED ISSUE #18717: gRPC cancellation of an in-progress call is broken in Blazor WebAssembly
_From @JamesNK on Wednesday, February 5, 2020 10:05:52 AM_

### Describe the bug
Canceling a gRPC-Web call that has returned headers but has not finished returning a response is broken in Blazor WebAssembly. Cancellation of a `HttpClient` request that has returned headers is done by calling `HttpResponseMessage.Dispose`. This should abort the HTTP request and set ASP.NET Core's `HttpContext.RequestAborted.IsCancellationRequested` to true. It appears that nothing happens currently.

Note that canceling a request that hasn't finished sending response headers (using the `SendAsync(request, ct)` cancellation token) does work.

### To Reproduce
I have a sample that attempts to do Blazor WA server streaming here. Clicking stop will cancel the call with the gRPC client. It calls `HttpResponseMessage.Dispose` internally

Source: https://github.com/JamesNK/grpc-dotnet/tree/jamesnk/blazor-serverstreaming (Blazor sample)

### Further technical details
- Blazor 3.2.0-preview1.20073.1

_Copied from original issue: dotnet/aspnetcore#18817_

### PR DESCRIPTION
**Pull request by campersau**
…nse content is disposed

Previously the fetch request was only aborted by the passed cancellation token to `SendAsync`. Now it is also cancelled when the `HttpResponseMessage`, `HttpContent` or the response `Stream` is disposed.

Best viewed without whitespace changes (`w=1`).

Fixes #18717 

---

How I tested it:
- Using example app from https://github.com/grpc/grpc-dotnet/tree/master/examples/Blazor
- Build `mono/sdks/wasm/framework/src/WebAssembly.Framework.sln`
- Copy DLLs from `mono/sdks/wasm/framework/netstandard2.0` into `C:\Users\[user]\.nuget\packages\microsoft.aspnetcore.blazor.mono\[version]\tools\mono\framework`
- Rebuild example app

### CODE REVIEW

File: sdks/wasm/framework/src/WebAssembly.Net.Http/WasmHttpMessageHandler.cs [hunk 1]
```diff
@@ -364,30 +362,40 @@ public override async Task<int> ReadAsync (byte [] buffer, int offset, int count
...
-					// value for fetch streams is a Uint8Array
-					using (Uint8Array binValue = (Uint8Array)read.GetObjectProperty ("value"))
-						_bufferedBytes = binValue.ToArray ();
+						_position = 0;
+						// value for fetch streams is a Uint8Array
+						using (Uint8Array binValue = (Uint8Array)read.GetObjectProperty("value"))
+							_bufferedBytes = binValue.ToArray ();
+					}
+				} catch (JSException) {
+					cancellationToken.ThrowIfCancellationRequested ();
```
Comment (campersau) [inline sdks/wasm/framework/src/WebAssembly.Net.Http/WasmHttpMessageHandler.cs]: When the request is aborted by disposing the response then don't throw `JSException` if the passed cancellation token is already cancelled.

File: sdks/wasm/framework/src/WebAssembly.Net.Http/WasmHttpMessageHandler.cs [hunk 2]
```diff
@@ -364,30 +362,40 @@ public override async Task<int> ReadAsync (byte [] buffer, int offset, int count
...
 
-				var t = (Task<object>)_reader.Invoke ("read");
-				using (var read = (JSObject)await t) {
-					if ((bool)read.GetObjectProperty ("done")) {
-						_reader.Dispose ();
-						_reader = null;
+				try {
+					var t = (Task<object>)_reader.Invoke ("read");
+					using (var read = (JSObject)await t) {
+						if ((bool)read.GetObjectProperty ("done")) {
```
Comment (campersau) [inline sdks/wasm/framework/src/WebAssembly.Net.Http/WasmHttpMessageHandler.cs]: Maybe check if passed cancellation token is cancelled after awaiting `read`? (Not required for this fix / PR.)

**Comment by lewing** [comment]
@monojenkins build

**Comment by lewing** [comment]
This sounds like something that should be included in the browser tests

**Comment by lewing** [comment]
@monojenkins build

**Comment by lewing** [comment]
@monojenkins build Linux AArch64

**Comment by lewing** [comment]
@kjpou1 I'd like to land this, can you please write a test then get it in?

**Comment by kjpou1** [comment]
Can recreate this reliably without Blazor code.  Test will follow in separate PR.

Thanks for the fix.

**Pull request merged**